pkgs/buildNpmPackage/mkcache.js

```diff
@@ -0,0 +1,38 @@
+const assert = require('assert')
+const fs     = require('fs')
+const path   = require('path')
+const {promisify} = require('util')
+
+// find pacote from npm dependencies
+module.paths.push(path.join(process.argv[0], "../../lib/node_modules/npm/node_modules"))
+const pacote = require('pacote')
+
+function traverseDeps(pkg, fn) {
+    Object.values(pkg.dependencies).forEach(dep => {
+        if (dep.resolved && dep.integrity) fn(dep)
+        if (dep.dependencies) traverseDeps(dep, fn)
+    })
+}
+
+async function main(lockfile, nix, cache) {
+    var promises = Object.keys(nix).map(async function(url) {
+        var tar = nix[url]
+        const manifest = await pacote.manifest(tar, {offline: true, cache})
+        return [url, manifest._integrity]
+    })
+    var hashes = new Map(await Promise.all(promises))
+    traverseDeps(lockfile, dep => {
+        if (dep.integrity.startsWith("sha1-")) {
+            assert(hashes.has(dep.resolved))
+            dep.integrity = hashes.get(dep.resolved)
+        }
+        else {
+            assert(dep.integrity == hashes.get(dep.resolved))
+        }
+    })
+    await promisify(fs.writeFile)("./package-lock.json", JSON.stringify(lock, null, 4))
```

Comment:
Are you just pretty-printing the same file here or am I missing something?

Reply:
This script updates the hashes that npm uses for the content-addressable cache from sha1 to sha512.

https://github.com/zkat/pacote/issues/162

Reply:
Yeah I realised that by now. Thanks.